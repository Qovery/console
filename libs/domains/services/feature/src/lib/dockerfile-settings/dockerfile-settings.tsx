import { type JobLifecycleTypeEnum } from 'qovery-typescript-axios'
import { type PropsWithChildren } from 'react'
import { Controller, type ControllerRenderProps, type UseFormReturn } from 'react-hook-form'
import { useParams } from 'react-router-dom'
import { match } from 'ts-pattern'
import {
  BlockContent,
  Button,
  Callout,
  CodeEditor,
  ExternalLink,
  Heading,
  Icon,
  InputSelect,
  InputText,
  Section,
  useModal,
} from '@qovery/shared/ui'
import { upperCaseFirstLetter } from '@qovery/shared/util-js'
import { DockerfileRawModal } from '../dockerfile-raw-modal/dockerfile-raw-modal'

type FileSource = 'GIT_REPOSITORY' | 'DOCKERFILE_RAW'

export interface DockerfileSettingsData {
  dockerfile_source: FileSource
  dockerfile_path?: string | null
  dockerfile_raw?: string | null
}

export interface DockerfileSettingsProps extends PropsWithChildren {
  methods: UseFormReturn<DockerfileSettingsData>
  onSubmit: () => void
  directSubmit?: boolean
  defaultContent?: string
  templateType?: JobLifecycleTypeEnum
}

export function DockerfileSettings({
  children,
  methods,
  onSubmit,
  directSubmit = false,
  defaultContent,
  templateType,
}: DockerfileSettingsProps) {
  const { environmentId = '' } = useParams()
  const { openModal, closeModal } = useModal()
  const { setValue, control, watch } = methods

  const availableFileSources: { label: string; value: FileSource }[] = [
    { label: 'Git repository', value: 'GIT_REPOSITORY' },
    { label: 'Raw Dockerfile', value: 'DOCKERFILE_RAW' },
  ]

  const watchDockerfileRaw = watch('dockerfile_raw')
  const watchDockerfileSource = watch('dockerfile_source')

  const handleSubmit = () => {
    if (watchDockerfileSource === 'GIT_REPOSITORY') {
      setValue('dockerfile_raw', null)
    } else {
      setValue('dockerfile_path', null)
    }
    onSubmit()
  }

  const openModalDockerfileRaw = (field: ControllerRenderProps<DockerfileSettingsData, 'dockerfile_raw'>) => {
    openModal({
      content: (
        <DockerfileRawModal
          environmentId={environmentId}
          content={field.value ?? ''}
          description={match(templateType)
            .with(
              'CLOUDFORMATION',
              'TERRAFORM',
              (templateType) =>
                `Qovery provides you with a default Dockerfile to deploy your ${upperCaseFirstLetter(templateType)}.`
            )
            .with('GENERIC', undefined, () => undefined)
            .exhaustive()}
          onClose={closeModal}
          onSubmit={(dockerfile_raw) => {
            field.onChange(dockerfile_raw)
            if (directSubmit) {
              handleSubmit()
            }
          }}
          defaultContent={defaultContent}
        />
      ),
      options: {
        fullScreen: true,
      },
    })
  }

  return (
    <form
      className="w-full space-y-10"
      onSubmit={(e) => {
        e.preventDefault()
        handleSubmit()
      }}
    >
      <Controller
        name="dockerfile_source"
        control={methods.control}
        defaultValue="GIT_REPOSITORY"
        render={({ field }) => (
          <InputSelect
            label="File source"
            value={field.value}
            onChange={field.onChange}
            options={availableFileSources}
          />
        )}
      />

      <Section className="gap-6">
        {watchDockerfileSource === 'DOCKERFILE_RAW' ? (
          <>
            <div>
              <Heading className="mb-2">Dockerfile as raw file</Heading>
              <p className="text-sm text-neutral-350">
                {match(templateType)
                  .with(
                    'CLOUDFORMATION',
                    'TERRAFORM',
                    (templateType) =>
                      `Customize here the dockerfile used to package your ${upperCaseFirstLetter(templateType)} and store it in the Qovery control plane.`
                  )
                  .with(
                    'GENERIC',
                    undefined,
                    () =>
                      'You can customize here the dockerfile to be used for this service and it will be stored on the Qovery control plane.'
                  )
                  .exhaustive()}
              </p>
            </div>

            {match(templateType)
              .with('CLOUDFORMATION', () => (
                <Callout.Root color="sky">
                  <Callout.Icon>
                    <Icon iconName="circle-info" iconStyle="regular" />
                  </Callout.Icon>
                  <Callout.Text>
                    Qovery provides a default Dockerfile and job parameters to deploy your template. These can be
                    customize based on your needs (additional arguments in the CLI commands etc..).
                  </Callout.Text>
                </Callout.Root>
              ))
              .with('TERRAFORM', () => (
                <Callout.Root color="sky">
                  <Callout.Icon>
                    <Icon iconName="circle-info" iconStyle="regular" />
                  </Callout.Icon>
                  <Callout.Text>
                    Qovery provides a default Dockerfile and job parameters to deploy your manifest. These can be
                    customize based on your needs (additional arguments in the CLI commands etc..).
                  </Callout.Text>
                </Callout.Root>
              ))
              .with('GENERIC', undefined, () => undefined)
              .exhaustive()}

            <Controller
              name="dockerfile_raw"
              control={control}
              rules={{
                required: 'Value required',
              }}
              render={({ field }) => (
                <BlockContent
                  title="Raw Dockerfile"
                  classNameContent="p-0"
                  headRight={
                    watchDockerfileRaw && (
                      <Button
                        aria-label="Edit"
                        type="button"
                        size="xs"
                        variant="surface"
                        onClick={() => openModalDockerfileRaw(field)}
                        className="hover:text-neutral-400"
                      >
                        <Icon iconName="pen" />
                      </Button>
                    )
                  }
                >
                  {watchDockerfileRaw ? (
                    <CodeEditor value={watchDockerfileRaw} language="dockerfile" readOnly height="300px" />
                  ) : (
                    <div className="my-4 px-10 py-5 text-center">
                      <Icon iconName="wave-pulse" className="text-neutral-350" />
                      <p className="mb-3 mt-1 text-xs font-medium text-neutral-350">No Dockerfile defined</p>
                      <Button type="button" size="md" onClick={() => openModalDockerfileRaw(field)}>
                        Create Dockerfile <Icon iconName="pen" className="ml-2" />
                      </Button>
                    </div>
                  )}
                </BlockContent>
              )}
            />
          </>
        ) : (
          <>
            <div>
              <Heading className="mb-2">Dockerfile from git repository</Heading>
              <p className="text-sm text-neutral-350">
                Specify the path containing the Dockerfile to be used to package your application.
              </p>
            </div>

            <Controller
              data-testid="input-text-dockerfile-path"
              key="dockerfile_path"
              name="dockerfile_path"
              defaultValue="Dockerfile"
              control={control}
              rules={{
                required: 'Value required',
              }}
              render={({ field, fieldState: { error } }) => (
                <InputText
                  dataTestId="input-text-dockerfile"
                  name={field.name}
                  onChange={field.onChange}
                  value={field.value ?? ''}
                  label="Dockerfile path"
                  error={error?.message}
                  hint={
                    <>
                      <span>Specify the location of your dockerfile. Expected format: myapp/Dockerfile</span>
                      <br />
                      <ExternalLink size="xs">Create one with Docker init</ExternalLink>
                    </>
                  }
                />
              )}
            />
          </>
        )}
      </Section>
      {directSubmit && watchDockerfileSource === 'DOCKERFILE_RAW' ? null : children}
    </form>
  )
}

export default DockerfileSettings
