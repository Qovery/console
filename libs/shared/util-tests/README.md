# shared-util-tests

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test shared-util-tests` to execute the unit tests via [Jest](https://jestjs.io).
