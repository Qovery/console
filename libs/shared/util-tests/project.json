{
  "name": "shared-util-tests",
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/shared/util-tests/src",
  "projectType": "library",
  "tags": ["scope:shared", "type:util"],
  "implicitDependencies": ["!shared-ui"],
  "targets": {
    "test": {
      "options": {
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "coverage": true
        }
      }
    }
  }
}
