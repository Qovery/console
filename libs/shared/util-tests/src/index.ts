export * from './lib/shared-util-tests'
