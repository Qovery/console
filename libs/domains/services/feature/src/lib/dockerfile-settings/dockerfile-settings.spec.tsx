import { useForm } from 'react-hook-form'
import selectEvent from 'react-select-event'
import { renderWithProviders, screen } from '@qovery/shared/util-tests'
import { DockerfileSettings, type DockerfileSettingsProps } from './dockerfile-settings'

// https://github.com/suren-atoyan/monaco-react/issues/88
// https://github.com/suren-atoyan/monaco-react/issues/88#issuecomment-887055307
jest.mock('@monaco-editor/react', () => {
  const { forwardRef } = jest.requireActual('react')
  const FakeEditor = forwardRef((props: any, ref: any) => (
    <textarea
      ref={ref}
      data-auto={props.wrapperClassName}
      onChange={(e) => props.onChange(e.target.value)}
      value={props.value}
    />
  ))
  return { Editor: FakeEditor }
})

type FormValues = {
  dockerfile_source: 'GIT_REPOSITORY' | 'DOCKERFILE_RAW'
  dockerfile_path: string | null
  dockerfile_raw: string | null
}

function DockerfileSettingsWithForm({
  onSubmit,
  ...props
}: Omit<DockerfileSettingsProps, 'methods' | 'onSubmit'> & {
  onSubmit: (data: FormValues) => void
}) {
  const form = useForm({
    mode: 'onChange',
    defaultValues: {
      dockerfile_source: 'GIT_REPOSITORY',
    },
  })
  const _onSubmit = form.handleSubmit((data) => {
    onSubmit(data as FormValues)
  })
  return <DockerfileSettings methods={form} onSubmit={_onSubmit} {...props} />
}

describe('DockerfileSettings', () => {
  it('should match snapshot without children', () => {
    const onSubmit = jest.fn()
    const { baseElement } = renderWithProviders(<DockerfileSettingsWithForm onSubmit={onSubmit} />)
    expect(baseElement).toMatchSnapshot()
  })

  it('should match snapshot with children', () => {
    const onSubmit = jest.fn()
    const { baseElement } = renderWithProviders(
      <DockerfileSettingsWithForm onSubmit={onSubmit}>
        <button type="submit">Submit</button>
      </DockerfileSettingsWithForm>
    )
    expect(baseElement).toMatchSnapshot()
  })

  it('should submit dockerfile path', async () => {
    const onSubmit = jest.fn()
    const { userEvent } = renderWithProviders(
      <DockerfileSettingsWithForm onSubmit={onSubmit}>
        <button type="submit">Submit</button>
      </DockerfileSettingsWithForm>
    )
    await userEvent.clear(screen.getByRole('textbox'))
    await userEvent.type(screen.getByRole('textbox'), 'CustomDockerfile')
    await userEvent.click(screen.getByRole('button', { name: /Submit/i }))
    expect(onSubmit).toHaveBeenCalledWith({
      dockerfile_source: 'GIT_REPOSITORY',
      dockerfile_raw: null,
      dockerfile_path: 'CustomDockerfile',
    })
  })

  it('should submit dockerfile raw', async () => {
    const onSubmit = jest.fn()
    const { userEvent } = renderWithProviders(
      <DockerfileSettingsWithForm onSubmit={onSubmit}>
        <button type="submit">Submit</button>
      </DockerfileSettingsWithForm>
    )
    const realSelect = screen.getByLabelText('File source')
    await selectEvent.select(realSelect, 'Raw Dockerfile')
    await userEvent.click(screen.getByRole('button', { name: /Create/i }))
    await userEvent.type(screen.getAllByRole('textbox')[0], 'my dockerfile content')
    await userEvent.click(screen.getByRole('button', { name: /Save/i }))
    await userEvent.click(screen.getByRole('button', { name: /Submit/i }))
    expect(onSubmit).toHaveBeenCalledWith({
      dockerfile_source: 'DOCKERFILE_RAW',
      dockerfile_path: null,
      dockerfile_raw: 'my dockerfile content',
    })
  })
})
