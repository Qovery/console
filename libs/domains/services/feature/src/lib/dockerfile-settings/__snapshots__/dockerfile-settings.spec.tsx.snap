// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DockerfileSettings should match snapshot with children 1`] = `
<body>
  <div>
    <form
      class="w-full space-y-10"
    >
      <div
        class=""
      >
        <div
          class="input input--select    "
          data-testid="select"
        >
          <label
            class="!translate-y-0 !text-xs"
            for="File source"
          >
            File source
          </label>
          <div
            class=" css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-3-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class="input-select__control css-13cymwt-control"
            >
              <div
                class="input-select__value-container input-select__value-container--has-value css-1fdsijx-ValueContainer"
              >
                <span
                  class="mr-1 text-sm text-neutral-400"
                >
                  Git repository
                </span>
                <input
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-readonly="true"
                  class="css-mohuvp-dummyInput-DummyInput"
                  id="File source"
                  inputmode="none"
                  role="combobox"
                  tabindex="0"
                  value=""
                />
              </div>
              <div
                class="input-select__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class="input-select__indicator-separator css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="input-select__indicator input-select__dropdown-indicator css-1xc3v61-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <input
              name="File source"
              type="hidden"
              value="GIT_REPOSITORY"
            />
          </div>
          <input
            name="File source"
            type="hidden"
            value="GIT_REPOSITORY"
          />
          <div
            class="pointer-events-none absolute right-4 top-1/2 -translate-y-1/2"
          >
            <span
              class="icon-solid-angle-down shrink-0 text-sm text-neutral-400"
              role="img"
            />
          </div>
        </div>
      </div>
      <section
        class="flex flex-col gap-6"
      >
        <div>
          <h1
            class="text-neutral-400 text-xl font-bold mb-2"
          >
            Dockerfile from git repository
          </h1>
          <p
            class="text-sm text-neutral-350"
          >
            Specify the path containing the Dockerfile to be used to package your application.
          </p>
        </div>
        <div
          class=""
          data-testid="input-text-dockerfile-wrapper"
        >
          <div
            class="relative"
          >
            <div
              aria-label="input-container"
              class="input    input--label-up "
            >
              <div
                class=""
              >
                <label
                  class="translate-y-2 text-sm"
                  for="Dockerfile path"
                >
                  Dockerfile path
                </label>
                <input
                  class="input__value "
                  data-testid="input-text-dockerfile"
                  id="Dockerfile path"
                  name="dockerfile_path"
                  type="text"
                  value="Dockerfile"
                />
              </div>
            </div>
          </div>
          <p
            class="mt-0.5 px-3 text-xs font-normal text-neutral-350 dark:text-neutral-50"
          >
            <span>
              Specify the location of your dockerfile. Expected format: myapp/Dockerfile
            </span>
            <br />
            <a
              class="cursor-pointer transition duration-100 font-medium inline-flex items-center gap-1 text-sky-500 hover:text-sky-600 text-xs"
              rel="noopener noreferrer"
              target="_blank"
            >
              Create one with Docker init
              <i
                aria-hidden="true"
                class="fa-solid fa-arrow-up-right-from-square text-2xs leading-4"
              />
            </a>
          </p>
        </div>
      </section>
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</body>
`;

exports[`DockerfileSettings should match snapshot without children 1`] = `
<body>
  <div>
    <form
      class="w-full space-y-10"
    >
      <div
        class=""
      >
        <div
          class="input input--select    "
          data-testid="select"
        >
          <label
            class="!translate-y-0 !text-xs"
            for="File source"
          >
            File source
          </label>
          <div
            class=" css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-2-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class="input-select__control css-13cymwt-control"
            >
              <div
                class="input-select__value-container input-select__value-container--has-value css-1fdsijx-ValueContainer"
              >
                <span
                  class="mr-1 text-sm text-neutral-400"
                >
                  Git repository
                </span>
                <input
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-readonly="true"
                  class="css-mohuvp-dummyInput-DummyInput"
                  id="File source"
                  inputmode="none"
                  role="combobox"
                  tabindex="0"
                  value=""
                />
              </div>
              <div
                class="input-select__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class="input-select__indicator-separator css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="input-select__indicator input-select__dropdown-indicator css-1xc3v61-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <input
              name="File source"
              type="hidden"
              value="GIT_REPOSITORY"
            />
          </div>
          <input
            name="File source"
            type="hidden"
            value="GIT_REPOSITORY"
          />
          <div
            class="pointer-events-none absolute right-4 top-1/2 -translate-y-1/2"
          >
            <span
              class="icon-solid-angle-down shrink-0 text-sm text-neutral-400"
              role="img"
            />
          </div>
        </div>
      </div>
      <section
        class="flex flex-col gap-6"
      >
        <div>
          <h1
            class="text-neutral-400 text-xl font-bold mb-2"
          >
            Dockerfile from git repository
          </h1>
          <p
            class="text-sm text-neutral-350"
          >
            Specify the path containing the Dockerfile to be used to package your application.
          </p>
        </div>
        <div
          class=""
          data-testid="input-text-dockerfile-wrapper"
        >
          <div
            class="relative"
          >
            <div
              aria-label="input-container"
              class="input    input--label-up "
            >
              <div
                class=""
              >
                <label
                  class="translate-y-2 text-sm"
                  for="Dockerfile path"
                >
                  Dockerfile path
                </label>
                <input
                  class="input__value "
                  data-testid="input-text-dockerfile"
                  id="Dockerfile path"
                  name="dockerfile_path"
                  type="text"
                  value="Dockerfile"
                />
              </div>
            </div>
          </div>
          <p
            class="mt-0.5 px-3 text-xs font-normal text-neutral-350 dark:text-neutral-50"
          >
            <span>
              Specify the location of your dockerfile. Expected format: myapp/Dockerfile
            </span>
            <br />
            <a
              class="cursor-pointer transition duration-100 font-medium inline-flex items-center gap-1 text-sky-500 hover:text-sky-600 text-xs"
              rel="noopener noreferrer"
              target="_blank"
            >
              Create one with Docker init
              <i
                aria-hidden="true"
                class="fa-solid fa-arrow-up-right-from-square text-2xs leading-4"
              />
            </a>
          </p>
        </div>
      </section>
    </form>
  </div>
</body>
`;
